[flutter tools] Don't return success if we trigger runZoned's error callback

jamesderlin opened the pull request:
## Description

Triggering `runZoned`/`runZonedGuarded`'s error callback does not
necessarily mean that the body stops executing. (Consequently, that
also means that the error callback can be triggered multiple times.)

This allowed some types of crashes to not be properly reported.  In
such cases we would undesirably continue along the success path,
which calls `exit()` and prematurely terminates the process before
the error path completes. (It'd be nice to stop calling `exit()`
entirely, but I digress.)

Fix this by checking if the zone's error callback has fired before
exiting with a success code.  This also prevents prematurely
terminating before the error callback can finish reporting the crash.

## Related Issues

https://github.com/flutter/flutter/issues/56406

## Tests

I added the following tests:

* I added a test to try to simulate the type of crash that would trigger `runZoned`'s error callback.  I verified that the test fails without the rest of my change and passes with it.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

<!-- Links -->
[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

jamesderlin committed:
[flutter tools] Don't return success if we trigger runZoned's error callback

Triggering `runZoned`/`runZonedGuarded`'s error callback does not
necessarily mean that the body stops executing. (Consequently, that
also means that the error callback can be triggered multiple times.)

This allowed some types of crashes to not be properly reported.  In
such cases we would undesirably continue along the success path,
which calls `exit()` and prematurely terminates the process before
the error path completes. (It'd be nice to stop calling `exit()`
entirely, but I digress.)

Fix this by checking if the zone's error callback has fired before
exiting with a success code.  This also prevents prematurely
terminating before the error callback can finish reporting the crash.

zanderso commented on packages/flutter_tools/test/general.shard/runner/runner_test.dart:
```diff
@@ -138,15 +191,28 @@ void main() {
+    Timer.run(() => throw error);
+
+    // Give the Timer time to fire.
+    await Future<void>.delayed(const Duration(milliseconds: 100));
```
It might work in this case, but it's generally not a good idea to have explicit timers like this in tests. There are two options. The first option is to use a `FakeAsync` block. I think that might be overkill here. Instead, it looks like you might be able to pass around some `Future`s to await and some `Completer`s to complete to achieve the execution order that you're trying to exercise.

jamesderlin replied:
I've replaced this one with a `Completer` to wait on.  I didn't think that particular case would be controversial since it involves a 0-delay timer and a non-zero delay timer, and I expect that timer callbacks would fire in order.

I really don't like the `Timer` in `SlowCrashReporter`.  The point of that is to make `runner.run`'s failure path take longer than the success path if they're racing.  I don't think there's anything along the success path that I can wait on; I can't add a `Completer` to the test-implementation of `exit()` since the fix is to *not* call `exit()` in this situation.

I could add an explicit, global, `@visibleForTesting`-only `Completer` to `runner.dart` that will be completed when we continue down the success path.  That also feels gross, but I suppose it's better than the hard-coded delay.  Is that acceptable?

zanderso reviewed (commented):
Good catch. Just a comment about the test.

jamesderlin committed:
Update with review feedback from zanderso

Remove hard-coded wait durations.

Also adjust some names.

zanderso commented on packages/flutter_tools/test/general.shard/runner/runner_test.dart:
```diff
@@ -138,15 +191,31 @@ void main() {
+    });
+
+    await completer.future;
+    return FlutterCommandResult.success();
```
Does anything go wrong if you do `runCompleted.complete()` here instead?

jamesderlin replied:
It seems to work, but it would be a bit more brittle.  If there asynchronous work ever occurs after `FlutterCommand.runCommand()` returns, the test could pass when it should fail.

zanderso replied:
Sorry, I'm not quite getting it. Could you spell this out for me a bit more? What's the sequence of events that would be problematic?

jamesderlin replied:
If `CrashingFlutterCommand.runCommand` completed `runCompleted` instead, the desired sequence of events would be:

1. `CrashingFlutterCommand.runCommand` generates an asynchronous error.
2. The `onError` callback from `runner.run`'s `runZoned` call fires.  We proceed down the crash reporting path. `WaitingCrashReporter` waits for `runCompleted`.
3. `CrashingFlutterCommand.runCommand` continues, completes `runCompleted`, and returns.
4. `runner.run` continues in its `try` block.
5. `WaitingCrashReporter` is unblocked and successfully reports the crash.

If execution yields between steps 3 and 4, then step 5 could run first.  If step 5 runs before step 4, a regression where the `try` block makes its own call to `exit()` would not be detected.

zanderso replied:
Okay, thanks. I understand the concern now. Other than a stray call in `analyze_continuously.dart`, the tool only calls `exit()` in `runner.dart` and `base/signals.dart`. We should guard against proliferating calls to exit with a more on-the-nose test, but not in this PR. For this PR, I think it would be best to move the `runCompleted` completer from `runner.dart` into the test.

zanderso replied:
I filed https://github.com/flutter/flutter/issues/59338

jamesderlin committed:
Merge branch 'master' of github.com:flutter/flutter into jamesderlin/premature-exit

jamesderlin committed:
Update with more review feedback from zanderso

Move the command Completer into the test to avoid polluting the tool
code.  Note that this weakens the test and might prevent it from
catching regressions.

jamesderlin committed:
Fix analysis error

zanderso reviewed (approved):
Thanks!
